core/src/lib/channel.h

```diff
@@ -0,0 +1,349 @@
+  in(const in&) = delete;
+  in& operator=(const in&) = delete;
+
+  template <typename... Args> bool emplace(Args... args)
+  {
+    if (did_close) { return false; }
+    if (auto handle = shared->write_lock()) {
+      handle->data().emplace_back(std::forward<Args>(args)...);
+      return true;
+    } else {
+      close();
+      return false;
+    }
+  }
+
+  template <typename... Args> bool try_emplace(Args... args)
+  {
+    if (did_close) { return false; }
+    auto result = shared->try_write_lock();
+    if (std::holds_alternative<failed_to_acquire_lock>(result)) {
+      return false;
+    } else if (std::holds_alternative<channel_closed>(result)) {
+      close();
+      return false;
+    } else {
+      std::get<typename queue<T>::handle>(result).data().emplace_back(
+          std::forward<Args>(args)...);
+      return true;
+    }
+  }
+
+  void close()
+  {
+    if (!did_close) {
+      shared->close_in();
+      did_close = true;
+    }
+  }
+
+  bool closed() const { return did_close; }
+
+  ~in()
+  {
+    if (shared) { close(); }
+  }
+};
+
+template <typename T> class out {
+  std::shared_ptr<queue<T>> shared;
+  std::vector<T> cache;
+  std::size_t cache_iter{0};
+  bool did_close{false};
+
+ public:
+  out(std::shared_ptr<queue<T>> shared) : shared{std::move(shared)} {}
+  out(out&&) = default;
+  out& operator=(out&&) = default;
+  out(const out&) = delete;
+  out& operator=(const out&) = delete;
+
+  std::optional<T> get()
+  {
+    if (did_close) { return std::nullopt; }
+    update_cache();
+
+    if (cache.size() > cache_iter) {
+      std::optional result
+          = std::make_optional<T>(std::move(cache[cache_iter++]));
+      return result;
+    } else {
+      return std::nullopt;
+    }
+  }
+
+  std::optional<T> try_get()
+  {
+    if (did_close) { return std::nullopt; }
+    try_update_cache();
+
+    if (cache.size() > cache_iter) {
+      std::optional result
+          = std::make_optional<T>(std::move(cache[cache_iter++]));
+      return result;
+    } else {
+      return std::nullopt;
+    }
+  }
+
+  void close()
+  {
+    if (!did_close) {
+      cache.clear();
+      cache_iter = 0;
+      shared->close_out();
+      did_close = true;
+    }
+  }
+
+  bool closed() const { return did_close; }
+
+  ~out()
+  {
+    if (shared) { close(); }
+  }
+
+ private:
+  void do_update_cache(std::vector<T>& data)
+  {
+    cache.clear();
+    cache_iter = 0;
+    std::swap(data, cache);
+  }
+
+  void update_cache()
+  {
+    if (cache_iter == cache.size()) {
+      if (auto handle = shared->read_lock()) {
+        do_update_cache(handle->data());
+      } else {
+        // this can only happen if the channel was closed.
+        close();
+      }
+    }
+  }
+
+  void try_update_cache()
+  {
+    if (cache_iter == cache.size()) {
+      auto result = shared->try_read_lock();
+      if (std::holds_alternative<failed_to_acquire_lock>(result)) {
+        // intentionally left empty
```

Comment:
who is the caller here ? do you mean the `[try_]get()` function or the caller of `[try_]get()` ?
* in case 1 i would argue, that `try_get()` does not need feedback. It should only care whether the cache is empty or not.
* in case 2, the caller can use `.closed()` to check whether the channel was alive or not.